File: VB/Serialization/MySerializationContext.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Reflection
Imports DevExpress.Utils
Imports DevExpress.Utils.Serializing
Imports DevExpress.Utils.Serializing.Helpers

Namespace Serialization
	Public Class MySerializationContext
		Inherits SerializationContext
		Private customProperties() As String

		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Function AllowProperty(ByVal helper As SerializeHelperBase, ByVal obj As Object, ByVal prop As PropertyDescriptor, ByVal options As OptionsLayoutBase, ByVal isSerializing As Boolean) As Boolean
			Dim myHelper As MyDeserializeHelper = TryCast(helper, MyDeserializeHelper)
			If myHelper Is Nothing OrElse Not(TypeOf myHelper.RootObject Is MyGridView) Then
				Return MyBase.AllowProperty(helper, obj, prop, options, isSerializing)
			End If

			Dim allow As Boolean = MyBase.AllowProperty(helper, obj, prop, options, isSerializing)
			If customProperties Is Nothing OrElse Array.IndexOf(customProperties, prop.Name) = -1 Then
				Return allow
			End If

			Dim propInfos() As PropertyInfo = obj.GetType().GetProperties(BindingFlags.Public Or BindingFlags.Instance)
			Dim [property] As PropertyInfo = Array.Find(propInfos, Function(info) info.Name = prop.Name)
			Return [property] IsNot Nothing
		End Function

		Protected Overrides Sub CustomGetSerializableProperties(ByVal obj As Object, ByVal pairsList As List(Of SerializablePropertyDescriptorPair), ByVal props As PropertyDescriptorCollection)
			MyBase.CustomGetSerializableProperties(obj, pairsList, props)
			If DeserializeHelper Is Nothing OrElse DeserializeHelper.RootObject Is Nothing OrElse Not(TypeOf DeserializeHelper.RootObject Is MyGridView) Then
				Return
			End If

			Dim args As New SerializeCustomPropertiesEventArgs(obj)
			CType(DeserializeHelper.RootObject, MyGridView).RaiseDeserializeCustomProperties(args)
			customProperties = args.PropertyNames
			If customProperties Is Nothing Then
				Return
			End If

			For Each customProperty As String In customProperties
				For i As Integer = 0 To pairsList.Count - 1
					If pairsList(i).Property.Name = customProperty Then
						Dim attribute As XtraSerializableProperty = TryCast(pairsList(i).Property.Attributes(GetType(XtraSerializableProperty)), XtraSerializableProperty)
						If attribute Is Nothing Then
							attribute = New XtraSerializableProperty()
						End If
						Dim prop As PropertyDescriptor = pairsList(i).Property

						pairsList.RemoveAt(i)
						pairsList.Insert(i, New SerializablePropertyDescriptorPair(prop, attribute))

						Exit For
					End If
				Next i
			Next customProperty
		End Sub

		Protected Overrides Overloads Function GetProperties(ByVal obj As Object, ByVal store As IXtraPropertyCollection) As PropertyDescriptorCollection
			Dim properties As PropertyDescriptorCollection = MyBase.GetProperties(obj, store)

			For Each storeItem As XtraPropertyInfo In store
				Dim args As New CustomPropertyValueEventArgs(obj, storeItem.Name)
				args.PropertyValue = storeItem.Value
				CType(DeserializeHelper.RootObject, MyGridView).RaiseGetCustomPropertyValue(args)
			Next storeItem

			Return properties
		End Function

		Private privateDeserializeHelper As MyDeserializeHelper
		Public Property DeserializeHelper() As MyDeserializeHelper
			Get
				Return privateDeserializeHelper
			End Get
			Set(ByVal value As MyDeserializeHelper)
				privateDeserializeHelper = value
			End Set
		End Property
	End Class
End Namespace

File: CS/Serialization/MySerializationContext.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Reflection;
using DevExpress.Utils;
using DevExpress.Utils.Serializing;
using DevExpress.Utils.Serializing.Helpers;

namespace Serialization
{
	public class MySerializationContext : SerializationContext
	{
		string[] customProperties;

		public MySerializationContext()
			: base()
		{
		}

		protected override bool AllowProperty(SerializeHelperBase helper, object obj, PropertyDescriptor prop, OptionsLayoutBase options, bool isSerializing)
		{
			MyDeserializeHelper myHelper = helper as MyDeserializeHelper;
			if ( myHelper == null || !(myHelper.RootObject is MyGridView) )
				return base.AllowProperty(helper, obj, prop, options, isSerializing);

			bool allow = base.AllowProperty(helper, obj, prop, options, isSerializing);
			if ( customProperties == null || Array.IndexOf(customProperties, prop.Name) == -1 )
				return allow;

			PropertyInfo[] propInfos = obj.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);
			PropertyInfo property = Array.Find(propInfos, info => info.Name == prop.Name);
			return property != null;
		}

		protected override void CustomGetSerializableProperties(object obj, List<SerializablePropertyDescriptorPair> pairsList, PropertyDescriptorCollection props)
		{
			base.CustomGetSerializableProperties(obj, pairsList, props);
			if ( DeserializeHelper == null || DeserializeHelper.RootObject == null || !(DeserializeHelper.RootObject is MyGridView) )
				return;

			SerializeCustomPropertiesEventArgs args = new SerializeCustomPropertiesEventArgs(obj);
			((MyGridView)DeserializeHelper.RootObject).RaiseDeserializeCustomProperties(args);
			customProperties = args.PropertyNames;
			if ( customProperties == null )
				return;

			foreach ( string customProperty in customProperties )
			{
				for ( int i = 0; i < pairsList.Count; i++ )
					if ( pairsList[i].Property.Name == customProperty )
					{
						XtraSerializableProperty attribute = pairsList[i].Property.Attributes[typeof(XtraSerializableProperty)] as XtraSerializableProperty;
						if ( attribute == null )
							attribute = new XtraSerializableProperty();
						PropertyDescriptor prop = pairsList[i].Property;

						pairsList.RemoveAt(i);
						pairsList.Insert(i, new SerializablePropertyDescriptorPair(prop, attribute));

						break;
					}
			}
		}

		protected override PropertyDescriptorCollection GetProperties(object obj, IXtraPropertyCollection store)
		{
			PropertyDescriptorCollection properties = base.GetProperties(obj, store);

			foreach ( XtraPropertyInfo storeItem in store )
			{
				CustomPropertyValueEventArgs args = new CustomPropertyValueEventArgs(obj, storeItem.Name);
				args.PropertyValue = storeItem.Value;
				((MyGridView)DeserializeHelper.RootObject).RaiseGetCustomPropertyValue(args);
			}

			return properties;
		}

		public MyDeserializeHelper DeserializeHelper
		{
			get;
			set;
		}
	}
}
